### PR TITLE
Specify the capacity for the StringBuilder to avoid excess copying an…

### PR DESCRIPTION
**Pull request by TruFire**
…d memory usage.

### CODE REVIEW

**Comment by gregghz** [comment]
It looks like you are doing one extra iteration/copy of args (`val stringArgs = args.map(_.toString)`) so you don't have to compute .toString for each element twice (once in the foldLeft for argsLength and once in the parts.zip.foreach block).

Since you will evaluate toString everytime you use this interpolator no matter what, you can chnage the definition of `toJsParameterValue` to override toString with a val:

```
implicit def toJsParameterValue[A: JsParameter](a: A): JsParameterValue = {
  new JsParameterValue {
    override val toString: String = implicitly[JsParameter[A]].asJsString(a)
  }
}
```

Once that is done you can get rid of stringArgs without worrying about calling asJsString multiple times:

```
val partsLength = parts.foldLeft(0) { (combined, current) =>
  combined + current.length
}
val argsLength = args.foldLeft(0) { (combined, current) =>
  combined + current.toString.length
}

val builder = new StringBuilder(partsLength + argsLength)
parts.zip(args).foreach { case (part, arg) =>
  builder ++= part
  builder ++= arg.toString
}
```

That should remove one more tiny bit of memory pressure.

**Comment by gregghz** [comment]
Assuming `s""` is memory efficient enough, to simplify this you could just defer back to scala's normal string interpolator since `js` isn't doing anything special except that it adds a layer of type checking:

```
  implicit class JsParameterStringContext(val stringContext: StringContext) extends AnyVal {
    def js(args: JsParameterValue*): JsInterpolatedString = {
      JsInterpolatedString(stringContext.s(args))
    }
  }
```

You can also just rip the implementation from here: https://github.com/scala/scala/blob/v2.11.11/src/library/scala/StringContext.scala#L119

I _think_ it's memory efficient enough for your purposes. It's effectively creating the StringBuilder with `new StringBuilder(parts(0))` which allocates space for `parts(0).length+16` chars. That might not quite meet your needs.

**Comment by TruFire** [comment]
`new StringBuilder(parts(0))` doesn't supply sufficient memory. This change is mostly for very large arguments, not the literal string parts.

**Comment by gregghz** [comment]
the zip is going to make a copy of parts and args too. If memory is really the only concern, it'd make more sense to iterate on all that imperatively.

File: src/main/scala/com/lucidchart/android/javascript/package.scala [hunk 1]
```diff
@@ -23,7 +29,7 @@ package object javascript {
 
   implicit def toJsParameterValue[A: JsParameter](a: A): JsParameterValue = {
     new JsParameterValue {
-      override def toString: String = implicitly[JsParameter[A]].asJsString(a)
+      override val toString: String = implicitly[JsParameter[A]].asJsString(a)
```
Comment (sriraamas) [inline src/main/scala/com/lucidchart/android/javascript/package.scala:32]: why val and not a def?

**Review by sriraamas — APPROVED** [review]

File: src/main/scala/com/lucidchart/android/javascript/package.scala (hunk 1 above)
Comment (TruFire) [reply src/main/scala/com/lucidchart/android/javascript/package.scala:32]: Gregg

Comment (gregghz) [reply src/main/scala/com/lucidchart/android/javascript/package.scala:32]: it's so you only evaluate the string once ever instead of everytime you call it

**Pull request merged**